runtime-transaction/src/instruction_details.rs

```diff
@@ -0,0 +1,199 @@
+            .requested_heap_size
+            .map_or(MIN_HEAP_FRAME_BYTES, |(_index, requested_heap_size)| {
+                requested_heap_size
+            })
+            .min(MAX_HEAP_FRAME_BYTES);
+
+        // Calculate compute unit limit
+        let compute_unit_limit = self
+            .compute_budget_instruction_details
+            .requested_compute_unit_limit
+            .map_or_else(
+                || {
+                    // NOTE: to match current behavior of:
+                    // num_non_compute_budget_instructions * DEFAULT
+                    self.builtin_instruction_details
+                        .count_builtin_instructions
+                        .saturating_add(
+                            self.builtin_instruction_details
+                                .count_non_builtin_instructions,
+                        )
+                        .saturating_sub(
+                            self.compute_budget_instruction_details
+                                .count_compute_budget_instructions,
+                        )
+                        .saturating_mul(DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT)
+                },
+                |(_index, requested_compute_unit_limit)| requested_compute_unit_limit,
+            )
+            .min(MAX_COMPUTE_UNIT_LIMIT);
+
+        let compute_unit_price = self
+            .compute_budget_instruction_details
+            .requested_compute_unit_price
+            .map_or(0, |(_index, requested_compute_unit_price)| {
+                requested_compute_unit_price
+            });
+
+        let loaded_accounts_bytes =
+            if let Some((_index, requested_loaded_accounts_data_size_limit)) = self
+                .compute_budget_instruction_details
+                .requested_loaded_accounts_data_size_limit
+            {
+                NonZeroU32::new(requested_loaded_accounts_data_size_limit)
+                    .ok_or(TransactionError::InvalidLoadedAccountsDataSizeLimit)?
+            } else {
+                MAX_LOADED_ACCOUNTS_DATA_SIZE_BYTES
+            }
+            .min(MAX_LOADED_ACCOUNTS_DATA_SIZE_BYTES);
+
+        Ok(ComputeBudgetLimits {
+            updated_heap_bytes,
+            compute_unit_limit,
+            compute_unit_price,
+            loaded_accounts_bytes,
+        })
+    }
+
+    pub fn try_from<'a>(
+        instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
+    ) -> Result<Self> {
+        let mut compute_budget_instruction_details = ComputeBudgetInstructionDetails::default();
+        let mut builtin_instruction_details = BuiltinInstructionDetails::default();
+
+        for (i, (program_id, instruction)) in instructions.enumerate() {
+            compute_budget_instruction_details.process_instruction(
+                i as u8,
+                program_id,
+                instruction,
+            )?;
+            builtin_instruction_details.process_instruction(program_id, instruction)?;
```

Comment:
It seems reasonable that we'll eventually add more of these for stuff like pre-compiles. Not sure if others I am not currently thinking of.

Although it'd increase the complexity of this code, we could probably save some hashing and pubkey comparisons by having some sort of return value.
If it is a compute budget, we don't even need to hash for the builtin check - we already know its' in the map. Similarly for the the pre-compile signatures.
Also if it's a compute budget ix, we don't need the signature instruction details to process the instruction at all.

Leave it as is for now, just thinking of how we might improve efficiency in the future.

Reply:
yeah, pre-compiled signatures are in the pipeline to be added, just not in this PR.